plaidbench.py

```diff
@@ -80,6 +78,99 @@ def has_plaid():
+        raise ValueError('The number of examples per epoch is not divisble by the batch size.')
+
+
+def train(x_train, y_train, epoch_size, model, batch_size, compile_stop_watch, epochs, stop_watch, output):
+    # Training
+    x = x_train[:epoch_size]
+    y = y_train[:epoch_size]
+    model.train_on_batch(x_train[0:batch_size], y_train[0:batch_size])
+    
+    compile_stop_watch.stop()
+
+    for i in range(epochs):
+        if i == 1:
+            printf('Doing the main timing')
+        stop_watch.start()
+        history = model.fit(
+            x=x, y=y, batch_size=batch_size, epochs=1, shuffle=False, initial_epoch=0)
+        stop_watch.stop()
+        time.sleep(.025 * random.random())
+        if i == 0:
+            output.contents = [history.history['loss']]
+    output.contents = np.array(output.contents)
+
+
+def inference(network, model, batch_size, compile_stop_watch, output, x_train, examples, stop_watch):
+    # Inference
+    y = model.predict(x=x_train, batch_size=batch_size)
+    
+    compile_stop_watch.stop()
+    output.contents = y
+    printf('Warmup')
+
+    for i in range(32 // batch_size + 1):
+        y = model.predict(x=x_train, batch_size=batch_size)
+
+    for i in range(examples // batch_size):
+        stop_watch.start()
+        y = model.predict(x=x_train, batch_size=batch_size)
+        stop_watch.stop()
+        time.sleep(.025 * random.random())
+
+
+def setup(train, epoch_size, batch_size):
+    # Setup
+    if train:
+        # Training setup
+        from keras.datasets import cifar10
+        from keras.utils.np_utils import to_categorical
+        printf('Loading the data')
+        (x_train, y_train_cats), (x_test, y_test_cats) = cifar10.load_data()
+        x_train = x_train[:epoch_size]
+        y_train_cats = y_train_cats[:epoch_size]
+        y_train = to_categorical(y_train_cats, num_classes=1000)
+    else:
+        # Inference setup
+        this_dir = os.path.dirname(os.path.abspath(__file__))
+        cifar_path = os.path.join(this_dir, 'cifar16.npy')
+        x_train = np.load(cifar_path).repeat(1 + batch_size // 16, axis=0)[:batch_size]
+        y_train_cats = None
+        y_train = None
+    return x_train, y_train
+
+
+def load_model(module, x_train):
+    this_dir = os.path.dirname(os.path.abspath(__file__))
+    module = os.path.join(this_dir, 'networks', '%s.py' % module)
+    globals = {}
+    exec_(open(module).read(), globals)
+    x_train = globals['scale_dataset'](x_train)
+    model = globals['build_model']()
+    printf("Model loaded.")
+    return module, x_train, model
+
+
+def run_intial(batch_size, compile_stop_watch, network, model):
+    print("Compiling and running initial batch, batch_size={}".format(batch_size))
+    compile_stop_watch.start()
+    optimizer = 'sgd'
+    if network[:3] == 'vgg':
+        from keras.optimizers import SGD
+        optimizer = SGD(lr=0.0001)
+    model.compile(optimizer=optimizer, loss='categorical_crossentropy',
```

Comment:
It's confusing, but 'compile' in keras doesn't actually compile tile code. If you read through the old code, you'll see that the compile stop watch measures one call to train and one call to infer before stopping the compile watch and starting the main one. It's impossible with the current way things are setup to isolate compilation time, so currently compile measures compilation + 1 batch of inference or training. You'll have to remove this function and just have the infer and train functions, and pass the compile stop watch to both.